File: spine-cpp/spine-cpp/src/spine/Skin.cpp
/******************************************************************************
* Spine Runtimes Software License v2.5
*
* Copyright (c) 2013-2016, Esoteric Software
* All rights reserved.
*
* You are granted a perpetual, non-exclusive, non-sublicensable, and
* non-transferable license to use, install, execute, and perform the Spine
* Runtimes software and derivative works solely for personal or internal
* use. Without the written permission of Esoteric Software (see Section 2 of
* the Spine Software License Agreement), you may not (a) modify, translate,
* adapt, or develop new applications using the Spine Runtimes or otherwise
* create derivative works or improvements of the Spine Runtimes or (b) remove,
* delete, alter, or obscure any trademarks or any copyright, trademark, patent,
* or other intellectual property or proprietary rights notices on or in the
* Software, including any copy thereof. Redistributions in binary or source
* form must include this license and terms.
*
* THIS SOFTWARE IS PROVIDED BY ESOTERIC SOFTWARE "AS IS" AND ANY EXPRESS OR
* IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
* MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
* EVENT SHALL ESOTERIC SOFTWARE BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
* SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
* PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES, BUSINESS INTERRUPTION, OR LOSS OF
* USE, DATA, OR PROFITS) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
* IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
* ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
* POSSIBILITY OF SUCH DAMAGE.
*****************************************************************************/

#include <spine/Skin.h>

#include <spine/Attachment.h>
#include <spine/Skeleton.h>

#include <spine/Slot.h>

#include <assert.h>

using namespace Spine;

Skin::AttachmentKey::AttachmentKey(int slotIndex, const char* name) :
		_slotIndex(slotIndex),
		_name(name, true) {
}

Skin::AttachmentKey::AttachmentKey(int slotIndex, const String &name) :
		_slotIndex(slotIndex),
		_name(name) {
}

bool Skin::AttachmentKey::operator==(const AttachmentKey &other) const {
	return _slotIndex == other._slotIndex && _name == other._name;
}

std::size_t Skin::HashAttachmentKey::operator()(const Spine::Skin::AttachmentKey &val) const {
	std::size_t h1 = val._slotIndex;
	return h1;
}

Skin::Skin(const String &name) : _name(name) {
	assert(_name.length() > 0);
}

Skin::~Skin() {
	HashMap<AttachmentKey, Attachment *>::Entries entries = _attachments.getEntries();
	while (entries.hasNext()) {
		HashMap<AttachmentKey, Attachment *>::Pair pair = entries.next();
		delete pair.value;
	}
}

void Skin::addAttachment(int slotIndex, const String &name, Attachment *attachment) {
	assert(attachment);
	_attachments.put(AttachmentKey(slotIndex, name), attachment);
}

Attachment *Skin::getAttachment(int slotIndex, const String &name) {
	AttachmentKey key(slotIndex, name.buffer());
	if (_attachments.containsKey(key)) {
		Attachment *attachment = _attachments[key];
		key.getName().unown();
		return attachment;
	} else {
		key.getName().unown();
		return NULL;
	}
}

void Skin::findNamesForSlot(int slotIndex, Vector<String> &names) {
	HashMap<AttachmentKey, Attachment *>::Entries entries = _attachments.getEntries();
	while (entries.hasNext()) {
		HashMap<AttachmentKey, Attachment *>::Pair pair = entries.next();
		if (pair.key._slotIndex == slotIndex) {
			names.add(pair.key._name);
		}
	}
}

void Skin::findAttachmentsForSlot(int slotIndex, Vector<Attachment *> &attachments) {
	HashMap<AttachmentKey, Attachment *>::Entries entries = _attachments.getEntries();
	while (entries.hasNext()) {
		HashMap<AttachmentKey, Attachment *>::Pair pair = entries.next();
		if (pair.key._slotIndex == slotIndex) {
			attachments.add(pair.value);
		}
	}
}

const String &Skin::getName() {
	return _name;
}

HashMap<Skin::AttachmentKey, Attachment *> &Skin::getAttachments() {
	return _attachments;
}

void Skin::attachAll(Skeleton &skeleton, Skin &oldSkin) {
	Vector<Slot *> &slots = skeleton.getSlots();
	HashMap<AttachmentKey, Attachment *>::Entries entries = oldSkin.getAttachments().getEntries();
	while (entries.hasNext()) {
		HashMap<AttachmentKey, Attachment *>::Pair pair = entries.next();
		int slotIndex = pair.key._slotIndex;
		Slot *slot = slots[slotIndex];

		if (slot->getAttachment() == pair.value) {
			Attachment *attachment = NULL;
			if ((attachment = getAttachment(slotIndex, pair.key._name))) {
				slot->setAttachment(attachment);
			}
		}
	}
}

File: spine-lua/spine-lua/Skin.lua
-------------------------------------------------------------------------------
-- Spine Runtimes License Agreement
-- Last updated January 1, 2020. Replaces all prior versions.
--
-- Copyright (c) 2013-2020, Esoteric Software LLC
--
-- Integration of the Spine Runtimes into software or otherwise creating
-- derivative works of the Spine Runtimes is permitted under the terms and
-- conditions of Section 2 of the Spine Editor License Agreement:
-- http://esotericsoftware.com/spine-editor-license
--
-- Otherwise, it is permitted to integrate the Spine Runtimes into software
-- or otherwise create derivative works of the Spine Runtimes (collectively,
-- "Products"), provided that each user of the Products must obtain their own
-- Spine Editor license and redistribution of the Products in any form must
-- include this license and copyright notice.
--
-- THE SPINE RUNTIMES ARE PROVIDED BY ESOTERIC SOFTWARE LLC "AS IS" AND ANY
-- EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
-- WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
-- DISCLAIMED. IN NO EVENT SHALL ESOTERIC SOFTWARE LLC BE LIABLE FOR ANY
-- DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
-- (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES,
-- BUSINESS INTERRUPTION, OR LOSS OF USE, DATA, OR PROFITS) HOWEVER CAUSED AND
-- ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
-- (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
-- THE SPINE RUNTIMES, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-------------------------------------------------------------------------------

local setmetatable = setmetatable
local table_insert = table.insert
local AttachmentType = require "spine-lua.attachments.AttachmentType"

local SkinEntry = {}
SkinEntry.__index = SkinEntry

function SkinEntry.new (slotIndex, name, attachment)
	local self = {
		slotIndex = slotIndex,
		name = name,
		attachment = attachment
	}
	setmetatable(self, SkinEntry)

	return self
end

local Skin = {}
Skin.__index = Skin

function Skin.new (name)
	if not name then error("name cannot be nil", 2) end

	local self = {
		name = name,
		attachments = {},
		bones = {},
		constraints = {}
	}
	setmetatable(self, Skin)

	return self
end

function Skin:setAttachment (slotIndex, name, attachment)
	if not name then error("name cannot be nil.", 2) end
	if not self.attachments[slotIndex] then self.attachments[slotIndex] = {} end
	self.attachments[slotIndex][name] = attachment
end

function Skin:addSkin (skin)
	for i, bone in ipairs(skin.bones) do
		local contained = false
		for j, otherBone in ipairs(self.bones) do
			if otherBone == bone then
			contained = true
			break
			end
		end
		if not contained then table_insert(self.bones, bone) end
	end

	for i, constraint in ipairs(skin.constraints) do
		local contained = false
		for j, otherConstraint in ipairs(self.constraints) do
			if otherConstraint == constraint then
				contained = true
				break
			end
		end
		if not contained then table_insert(self.constraints, constraint) end
	end

	local attachments = skin:getAttachments()
	for i, entry in ipairs(attachments) do
		self:setAttachment(entry.slotIndex, entry.name, entry.attachment)
	end
end

function Skin:copySkin (skin)
	for i, bone in ipairs(skin.bones) do
		local contained = false
		for j, otherBone in ipairs(self.bones) do
			if otherBone == bone then
			contained = true
			break
			end
		end
		if not contained then table_insert(self.bones, bone) end
	end

	for i, constraint in ipairs(skin.constraints) do
		local contained = false
		for j, otherConstraint in ipairs(self.constraints) do
			if otherConstraint == constraint then
			contained = true
			break
			end
		end
		if not contained then table_insert(self.constraints, constraint) end
	end

	local attachments = skin:getAttachments()
	for i, entry in ipairs(attachments) do
		if entry.attachment.type == AttachmentType.mesh then
			entry.attachment = entry.attachment:newLinkedMesh()
			self:setAttachment(entry.slotIndex, entry.name, entry.attachment)
		else
			entry.attachment = entry.attachment:copy()
			self:setAttachment(entry.slotIndex, entry.name, entry.attachment)
		end
	end
end

function Skin:getAttachment (slotIndex, name)
	if not name then error("name cannot be nil.", 2) end
	local dictionary = self.attachments[slotIndex]
	if dictionary then
		return dictionary[name]
	else
		return nil
	end
end

function Skin:removeAttachment (slotIndex, name)
	local slotAttachments = self.attachments[slotIndex]
	if slotAttachments then
		slotAttachments[name] = nil
	end
end

function Skin:getAttachments ()
	local entries = {}
	for slotIndex, slotAttachments in pairs(self.attachments) do
		if slotAttachments then
			for name, attachment in pairs(slotAttachments) do
				if attachment then
					table_insert(entries, SkinEntry.new(slotIndex, name, attachment))
				end
			end
		end
	end
	return entries
end

function Skin:getAttachmentsForSlot (slotIndex)
	local entries = {}
	local slotAttachments = self.attachments[slotIndex]
	if slotAttachments then
		for name, attachment in pairs(slotAttachments) do
			if attachment then
				table_insert(entries, SkinEntry.new(slotIndex, name, attachment))
			end
		end
	end
	return entries
end

function Skin:clear ()
	self.attachments = {}
	self.bones = {}
	self.constraints = {}
end

function Skin:attachAll(skeleton, oldSkin)
	for i, slot in ipairs(skeleton.slots) do
		local slotAttachment = slot.attachment
		if slotAttachment then
			local dictionary = oldSkin.attachments[i]
			if (dictionary) then
				for key, value in pairs(dictionary) do
					local skinAttachment = value
					if slotAttachment == skinAttachment then
						local attachment = self:getAttachment(i, key)
						if attachment then
							slot:setAttachment(attachment)
						end
						break
					end
				end
			end
		end
	end
end

function Skin:findNamesForSlot (slotIndex)
	local names = {}
	for _,v in self.attachments do
		if v[1] == slotIndex then table_insert(names, v[2]) end
	end
end

function Skin:findAttachmentsForSlot (slotIndex)
	local attachments = {}
	for _,v in self.attachments do
		if v[1] == slotIndex then table_insert(attachments, v[3]) end
	end
end

return Skin
